MINOR: Use readable interface to parse response

soondenana opened the pull request:
The generated response data classes take Readable as input to parse the
Response. However, the associated response objects take ByteBuffer as
input and thus convert them to Readable using `new ByteBufferAccessor`
call.

This PR changes the parse method of all the response classes to take the
Readable interface instead so that no such conversion is needed.

To support parsing the ApiVersionsResponse twice for different version
this change adds the "slice" method to the Readable interface.

Reviewers: José Armando García Sancio <jsancio@apache.org>, Truc Nguyen
 <[trnguyen@confluent.io](mailto:trnguyen@confluent.io)>, Aadithya
 Chandra <[aadithya.c@gmail.com](mailto:aadithya.c@gmail.com)>

soondenana committed:
MINOR: Use readable interface to parse response

This is the Response side of Request  refactoring PR in commit: 56d1dc1b6e12a8a325c91749485aa03bce17c7f1

The generated ReequestData classes take Readable as input to parse the Response. However,
the individual Response objects take ByteBuffer as input and thus convert them to Readable
using `new ByteBufferAccessor` call. This call is in all the Respose classes that create
boilderplate code in high tens of these classes.

This PR changes the parse method of all the Respnose class to take Readable instead so that
no such conversion is needed. The clients are the responsible for calling the parse method
with appropriate object which is majority of the cases is already a Readable.

There were couple of places in code where `position` method is used, which is not present
in the Readable interface, so a new `ReadableBuf` interface is created to add these methods.
This allows us to not retreive ByteBuffer from Readable to use the `position` method.

Rest of the changes are all refactoring with no logical change.

trnguyencflt commented:
[nit] Please remove this block "Delete this text ...." from the description of the PR.

Otherwise, LGTM

trnguyencflt commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java:
```diff
@@ -51,8 +53,8 @@ final ByteBuffer serializeWithHeader(ResponseHeader header, short version) {
     // Visible for testing
-    final ByteBuffer serialize(short version) {
-        return MessageUtil.toByteBufferAccessor(data(), version).buffer();
+    final ByteBufferAccessor serialize(short version) {
```
should we return the interface (Readable/ReadableBuf) instead?

soondenana replied:
It was like that in the Request PR, but as feedback I was asked to change that, so kept it same in this PR.

ditac commented on clients/src/main/java/org/apache/kafka/common/protocol/ReadableBuf.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * A readable backed by some buffers
+ */
+public interface ReadableBuf extends Readable {
```
Why not make this part of Readable?

soondenana replied:
Thought about that when create the `ReadableBuf` interface, `Readable` only has methods related to reading byte/int/long/.... etc

It has no method to set things (like position) and also that feels more like a Buffer abstraction. Only method that recently got added to it was `remaining` which I would argue should be added the way this PR has, create an abstraction for a Readable that is backed by a buffer.

However, as there is only one implementation of it, so it doesn't matter and I will be happy to move it in `Readable`. Let me know.

soondenana committed:
Fix missing buffer call in RequestResponseTest

jsancio commented:
Thanks for the change @soondenana .

There are conflicts with the PR. The PR description doesn't match the code changes. For example, I cannot find this new ReadableBuf interface.

soondenana commented:
> Thanks for the change @soondenana .
> 
> There are conflicts with the PR. The PR description doesn't match the code changes. For example, I cannot find this new ReadableBuf interface.

Rebased, couple of parse methods (in FetchResponse and ShareFetchResponse) had comments added, which caused conflict. 

Also, updated the commit description. Original version of the commit I had tried to solve the issue by adding a new interface, but that does more than just "refactoring" so took that out and went with simpler approach. However, missed updating the comment in commit for that. Its now fixed.

jsancio commented on clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java:
```diff
@@ -152,19 +151,19 @@ public boolean zkMigrationReady() {
     }
 
-    public static ApiVersionsResponse parse(ByteBuffer buffer, short version) {
+    public static ApiVersionsResponse parse(ByteBufferAccessor readable, short version) {
```
The rest of the implementations of `parse` but this one take a `Readable`. Did you consider adding `Readable slice()` to `Readable`? If you do that, this implementation can call `slice()`so that it can parse the response twice if the version is incorrect.

soondenana replied:
Done.

jsancio commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java:
```diff
@@ -106,189 +107,189 @@ public static AbstractResponse parseResponse(ByteBuffer buffer, RequestHeader re
     }
 
-    public static AbstractResponse parseResponse(ApiKeys apiKey, ByteBuffer responseBuffer, short version) {
+    public static AbstractResponse parseResponse(ApiKeys apiKey, ByteBufferAccessor readable, short version) {
```
If you agree with my suggestion, you can change this signature to `parseResponse(ApiKeys, Readable, short)`.

soondenana replied:
Done.

jsancio reviewed (commented):
Thanks for the changes. I left a comment on how you can change the code so that we can use `Readable` through out.

soondenana committed:
Added slice method to Readable and used it

soondenana committed:
Fix typo

jsancio commented on clients/src/main/java/org/apache/kafka/common/protocol/Readable.java:
```diff
@@ -38,6 +38,7 @@ public interface Readable {
     int readVarint();
     long readVarlong();
     int remaining();
+    Readable slice();
```
Can we add Java Doc comment to this new method. `Readable` is an important type and it should be fully documented but I understand that is may be out of scope of this PR.

soondenana replied:
Done

jsancio commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java:
```diff
@@ -110,7 +111,7 @@ public static AbstractResponse parseResponse(ByteBuffer buffer, RequestHeader re
     }
 
-    public static AbstractResponse parseResponse(ApiKeys apiKey, ByteBufferAccessor readable, short version) {
+    public static AbstractResponse parseResponse(ApiKeys apiKey, Readable readable, short version) {
```
Why don't you also fix the static method `AbstractResponse parseResponse(ByteBuffer, RequestHeader)`?

Are you planning to fix `RequestHeader` and `ResponseHeader` to use `Readable` instead of `ByteBuffer`?

soondenana committed:
Add javadoc

jsancio reviewed (approved):
Thanks for the changes @soondenana . It is good to see more code using the `Readable` interface.
